### ISSUE TITLE
New Task: Conversion to BUFR documentation/code

### ISSUE THREAD

**Issue by MathewBiddle** [ioos/marine_life_data_network#17]
### Who is requesting this?

@ioos/marine-life

### What is being requested?

Conversion to BUFR documentation/code.

### What is the requested deadline and why?

2024-07-01

### What is the current status quo (i.e., what happens if this does not get done)?

Lack of understanding if something breaks in the process to send data to the GTS. 

### Provide a description or any other important information.

What indicates this is done?

- [ ]  A pull request to https://github.com/ioos/ioos-atn-data/ with documentation of the process.
- [ ] The source code is published on GitHub at https://github.com/ioos/ioos-atn-data/ and well documented, or other appropriate repository.

@mmckinzie please elaborate on what is meant by 'code' above and include a statement as to what indicates that this is successful.

**Comment by MathewBiddle** [comment]
also, what are we converting from?

**Comment by mmckinzie** [comment]
Consider updating issue to: Conversion of animal borne profile data to BUFR for submission to the GTS

What is being requested?
- (Axiom) Conversion of TD and CTD data from animal borne platforms to BUFR messages, follow updated v.39 tables – [doc_num.php (wmo.int)](https://library.wmo.int/doc_num.php?explnum_id=11283); Right now for SMRU data only (CTD-SRDL, CTD/F-SRDL, GPS-SRDL), but eventually we’ll need to add the ability to serve data from additional manufacturers and tag types (e.g., WC SCOUT, Ornitela OrniTrack GPS/GSM tags)
- (Axiom) Successful submission of BUFR messages to NDBC staging platform for submission to the GTS in near r/t – Index of / (axds.co) NOTE: This ideally happens within 24 hours of profile generation/data collection by tag, but within a maximum of 10 days
- (Axiom) Troubleshoot and fix any issues preventing near r/t data submission to the GTS, as needed
- (Axiom) Update BUFR QC protocol, stop hardcoding GTS QC flags to 0, implement QARTOD to GTSPP flag conversion and add appropriate flags to BUFR messages, see related task under issue #14
- (ML) Submission of required/recommended OceanOPS metadata to the AniBOS OCG ERDDAP – ERDDAP - AniBOS Platform Metadata - Data Access Form (noaa.gov)
- (ML) Track ATN BUFR message submission to the GTS and integration in IOOS by numbers metric – [ioos_metrics/gts/GTS_ATN_monthly_totals.csv at main · ioos/ioos_metrics (github.com)](https://github.com/ioos/ioos_metrics/blob/main/gts/GTS_ATN_monthly_totals.csv) and  [IOOS By The Numbers - The U.S. Integrated Ocean Observing System (IOOS) (noaa.gov)](https://ioos.noaa.gov/about/ioos-by-the-numbers/)
- (Axiom) Link BUFR tools to atn data github – https://github.com/axiom-data-science/bufrtools
- (Axiom) publish code, including any modifications and documentation of ATN processing steps for conversion of these animal borne platforms sensor data to BUFR on GitHub (Need specific documentation by manufacturer?) 

Deadline: Conversion of sensor data to BUFR and submission of data to GTS and metadata to OCG ERDDAP is on-going, troubleshoot as needed; calculating IOOS by number metric is on-going; publish (link) code and documentation about ATN processing and BUFR message generation on atn data github – November 30, 2024

**Comment by MathewBiddle** [comment]
Thanks, @mmckinzie. This is a great summary and provides some great background. I think the majority of the intent for this issue is already done. You've done a great job documenting what the process looks like and who's responsible for what. Now we just need to write that down somewhere a little bit more permanent. I suggest a new page under https://ioos.github.io/ioos-atn-data/index.html. (I've added a new section for ATN Data Flows which this could fit into).

> Consider updating issue to: Conversion of animal borne profile data to BUFR for submission to the GTS

As this is an ongoing task, I think we should go the opposite way and make this issue more specific. Something like "for SMRU data, publish (link) code and documentation about ATN processing and BUFR message generation on https://ioos.github.io/ioos-atn-data/index.html" That way we understand what the existing process is, before moving to other activities.

Then, maybe next year (or whenever there is open space to develop or when we deem it a priority) we add the additional tag types as new projects.

> eventually we’ll need to add the ability to serve data from additional manufacturers and tag types (e.g., WC SCOUT, Ornitela OrniTrack GPS/GSM tags)

Ensuring we are linking to https://github.com/axiom-data-science/bufrtools will be crucial. We might need to think about making a fork of that under IOOS.

**Comment by MathewBiddle** [comment]
I would also like to know what NDBC does on their side. Do they do any validation of the BUFR messages? Does ATN receive a report back?

**Comment by mmckinzie** [comment]
No NDBC does not do any validation or QC of the messages, they accept IOOS ATN BUFR messages (TM315023) served to NDBC via our [staging platform](https://stage-ndbc-bufr.srv.axds.co/), adds the WMO header (IOXX01 KWBN) to create a GTS message which is delivered to NWSTG for distribution to the GTS. NDBC is essentially a pass-through mechanism.

**Comment by conniekot** [comment]
@iwensu0313 Please update this issue with progress, if it has been incorporated into https://github.com/ioos/ioos_qc, and any links to the code - thanks!

**Comment by iwensu0313** [comment]
Before I can provide an update, I need clarity on what we have agreed the scope of this issue is. Given the conversation here, I can't tell if the ticket description is up to date

> As this is an ongoing task, I think we should go the opposite way and make this issue more specific. Something like "for SMRU data, publish (link) code and documentation about ATN processing and BUFR message generation on https://ioos.github.io/ioos-atn-data/index.html" That way we understand what the existing process is, before moving to other activities.

Based on this most recent note from @MathewBiddle, it sounds like we want this task to focus on publishing code and documentation about conversion of netCDF files to BUFR? 

I know we have this published conversion code https://github.com/axiom-data-science/bufrtools. What are remaining tasks needed?

As Matt and Megan indicated above, both SMRU and WC profile netcdf data have been converted to BUFR format and been served [here](https://stage-ndbc-bufr.srv.axds.co/platforms/atn/) for NDBC to pick up and distribute to GTS for some years now. The pipeline currently is running smoothly. There was one issue we resolved this year related to WMO id integer overflow causing problems in the pipeline. That has been fixed - I believe we were able to fix it in our internal pipeline that imports `bufrtools` (a smaller lift than having to update `bufrtools` itself).

**Comment by iwensu0313** [comment]
In Megan's requested items for Axiom [overview](https://github.com/ioos/marine_life_data_network/issues/17#issuecomment-2081658068) above, I think this one still has open items

### (Axiom) Update BUFR QC protocol, stop hardcoding GTS QC flags to 0, implement QARTOD to GTSPP flag conversion and add appropriate flags to BUFR messages, see related task under issue https://github.com/ioos/marine_life_data_network/issues/14

#### stop hardcoding GTS QC flags to 0
* [PR](https://github.com/axiom-data-science/bufrtools/pull/12) here
* trello [here](https://trello.com/c/XO6gVizX)
* my notes as well as the trello indicates that we have been blocked on implementation until PIs review the PR 

From Stacey in Trello (June 2023):
> "In communicating with the BUFR folks, they should review the MR in the BUFRtools package to ensure our interpretation of the QARTOD to WMO flag mapping is correct."

### (Axiom) Link BUFR tools to atn data github – https://github.com/axiom-data-science/bufrtools
Do we still need to do something here?

**Comment by MathewBiddle** [comment]
Why did this get moved to blocked?

I think it would be helpful to document the BUFR process as a sub-page on https://ioos.github.io/ioos-atn-data/ under ATN Data Flows:
![Image](https://github.com/user-attachments/assets/58d8e028-316a-437c-933d-a6a435c7fe80)

That page would include an overarching diagram of the bufr conversion process and links to https://github.com/axiom-data-science/bufrtools. 

We need a one-stop-shop on what processing is done to the data received at ATN. I'd like that to be https://ioos.github.io/ioos-atn-data/index.html

**Comment by iwensu0313** [comment]
Moved to Blocked, unless we decide it is a priority for the upcoming 3 months (Dec 2024 - Feb 2025). Though a status like Paused would be more reflective. I wasn't sure!

**Documentation**
Okay, gotcha - a diagram makes sense! Here's a run-down for starters of what's happening:
- prepare animal borne profile data: 1) make sure depths are ordered by time 2) make sure profile id is 8 characters max
- the outputs of the following are submitted
    - Create a BUFR message for this profile by running `bufrtools.encoding.wildlife_computers.encode`: `encode(profile_dataset=df, output="profiles.bufr", uuid=random_uuid, ptt=random_ptt_id))` or from the command line as outlined in the bufrtools github readme
    - I think we also get the Section 4 Records of the BUFR Message by running `bufrtools.encoding.wildlife_computers.get_section4`: `records = get_section4(df, random_uuid)`

the conversion itself is all captured in https://github.com/axiom-data-science/bufrtools. are you thinking the steps within `bufrtools` should be translated into a flow chart as well that would be more human-readable, tells a story? Feels like that should go into the README of `bufrtools` rather than somewhere external. But we can add that as a to do if helpful.

a jupyter notebook _could_ work, but it's not exactly reproducing what we're doing since our systems are tightly interconnected (e.g. `uuid` is an identifier for each dataset in our system). And I don't know if it's all that useful. It depends if the goal is so that people can learn to use `bufrtools`, but that should just be documented in the README which it already is (there's a one-line command tool)

**Questions:**
> Not to track bugs/progress on the development of those processes. If that makes sense.

Yep - is it helpful to create new tickets for bugs as they come up? Or is that too detailed for this space?

Should we rename the title to "Document the Conversion of animal borne profile data to BUFR for submission to the GTS"

**Comment by MathewBiddle** [comment]
**Documentation**
Yes! That run-down is what I am looking for. Something in plain language that we can share with folks who might not fully understand all the nuances with ATN but can gather an understanding of what the data flow looks like. I didn't know about bufrtools and that it was the main package ATN is using, so connecting ATN with `bufrtools` in the documentation will be key.

> are you thinking the steps within `bufrtools` should be translated into a flow chart as well that would be more human-readable, tells a story?

Exactly. If putting it in the README.md makes more sense, that's fine. But, we need to connect ATN to `bufrtools` somewhere.

> a jupyter notebook could work, but it's not exactly reproducing what we're doing since our systems are tightly interconnected (e.g. uuid is an identifier for each dataset in our system).

Understood. That's fine with me. I was just looking for some examples of how `bufrtools` is being used at ATN.

> Yep - is it helpful to create new tickets for bugs as they come up? Or is that too detailed for this space?

Yes, as issues arise we need to know about them and evaluate if we have the resources to tackle them.

> Should we rename the title to "Document the Conversion of animal borne profile data to BUFR for submission to the GTS"

Yes, probably a good idea.

**Comment by iwensu0313** [comment]
Moved from Blocked to In Progress to pick back up where I left off. Discussed at Dec 2025 Marine Life kickoff that documentation is an immediate priority.

I'll start looking at posting a PR with the information requested in https://ioos.github.io/ioos-atn-data/

**Comment by iwensu0313** [comment]
Submitted PR https://github.com/ioos/ioos-atn-data/pull/112, would like Megan to do a sanity check review before we publish. Pinged her in the PR

**Comment by iwensu0313** [comment]
This PR has been reviewed by Megan and ready to be merged https://github.com/ioos/ioos-atn-data/pull/112. Once merged, this task can be closed.

**Comment by MathewBiddle** [comment]
Thanks @iwensu0313 and @mmckinzie for pulling this together! The content looks great!

**Comment by MathewBiddle** [comment]
page is published at https://ioos.github.io/ioos-atn-data/ndbc-gts.html
